### DISCUSSION TITLE
macOS Wine version working

### DISCUSSION THREAD

**Discussion by erichelgeson** [fadden/CiderPress2#34] [Show and tell]
The trick is to make sure you're using a 32 bit release of Wine, specifically I am using `WineCX21.2.0 32-Bit (wine-6.0.2)` and Windows 7 along with the bundled version of CiderPress. 

Seems to work great - I'm able to add files, menus work, nothing out of the ordinary besides the initial startup is a few seconds.

I know a native Mac GUI is not really on the roadmap as of yet, but this seems an OK way to use CiderPress2!

https://github.com/user-attachments/assets/532e1d08-ffb8-4eb9-8394-a8d3b6ebd1c2

I have a pre-bundled (unsigned, unnotarized) `CiderPress2.app` [here](https://mega.nz/folder/8hA3AQCJ#pWUq92L70yDXlogy9lk5Dg) in the "CiderPress2-Mac" folder.

Thanks again for the work you do!

**Comment by fadden** [comment]
That's excellent!  I'm amazed the drag & drop stuff works.

I didn't have nearly this level of success on my Linux system, but I suspect it's due to something with fonts (the crashes I've seen for both this and SourceGen are in the font code... and all the text looks really ugly).

**Reply by hippietrail** [reply]
I think the ugly text is due to fonts for properietary OSes using a type of hinting that is or was patented where Linux uses autohinting from memory and typically uses different fonts that don't rely on the patented hinting. Some fonts really overly rely on or abuse the patented hinting system so they look awful when it's not there.
I'm pretty sure the patents have now expired but Linux folks are used to using their own fonts and haven't implemented the patented hinting and haven't made it a priority as it only shows up when you use fonts from Windows and macOS on Linux.

**Reply by HelixNRG** [reply]
Hiya! The potential for using CiderPress II on the Mac is awesome, but I've tried downloading the provided WINE bottled CiderPress app from Mega, (I've tried downloading with Safari and Chrome, and extracted the .zip file with both macOS's built-in Zip utility and Keka as an alternative, but I always get the dialog saying the application is damaged and can't be run.

I'm using macOS 14.6.1 (23G93) on an M2 Pro MacBook Pro with 32GB RAM. Please help :-) 

<img width="1120" alt="image" src="https://github.com/user-attachments/assets/f03f84d1-8264-407f-ac28-adc188f2bc6b">

**Reply by fadden** [reply]
You're running into the macOS quarantine system.  Files downloaded from the web, or unzipped from files downloaded from the web, inherit a "quarantine" extended attribute.  For some reason, macOS declares that they are damaged rather than unsafe.  The native cp2 command has the same issue when downloaded.

To fix this, start a Terminal window, `cd` into the CiderPress-Mac folder, and run this command: `find . -print0 | xargs -0 xattr -d com.apple.quarantine`  [edited]

This recursively removes the quarantine attribute from all files and directories.  After you do this, it should work.  (Tested just now on macOS 11.6.3 running in emulation.)

**Reply by HelixNRG** [reply]
Thanks Andy! I've used Terminal and in the directory for the CiderPress2.app, but trying:

`find . -print0 | xargs -0 attr -d com.apple.quarantine`

results in:

`xargs: attr: No such file or directory`

I'm guessing (not regularly using the Terminal) that I'm missing some Command Line app to do this? Or is it another reason?

**Reply by fadden** [reply]
The problem is that I can't type.  The command should be `xattr`, not `attr`.  (I've edited the previous message to be correct.)

**Reply by HelixNRG** [reply]
Success! Thanks again!

<img width="792" alt="image" src="https://github.com/user-attachments/assets/443ea197-880a-4a67-b497-0ee427488bd8">


Look forward to playing it with it over time, but my time with retro computing is more limited than ever these days.

**Reply by hippietrail** [reply]
> Hiya! The potential for using CiderPress II on the Mac is awesome, but I've tried downloading the provided WINE bottled CiderPress app from Mega, (I've tried downloading with Safari and Chrome, and extracted the .zip file with both macOS's built-in Zip utility and Keka as an alternative, but I always get the dialog saying the application is damaged and can't be run.

Very annoyingly, Apple decided to invent a new meaning for "damaged" that causes confusion even in pretty tech-savvy people. I think they were trying too hard to avoid scary or technical language like "unsafe" or "quarantine" but they elected to mislead instead and give the impression something went wrong in the downloading or decompressing process.

**Reply by erichelgeson** [reply]
I've added these instructions to the readme in the zip, thanks!

**Comment by fadden** [comment]
I spent some time playing with this.  I wrote up my notes and some step-by-step instructions for Kegworks; you can find them here: https://github.com/fadden/CiderPress2/blob/main/WineNotes.md

The only difference from what you did is that I used Winetricks to install the Consolas font.  This makes the fixed-width font stuff look much better, e.g. the block editor is no longer a horror show.  I couldn't figure out how to add an arbitrary font though... would be nice to see how it looks with Segoe UI installed.

I tried to upload a wrapped ZIP of v1.0.6 to Google Drive, but the automated system flagged it as potential malware and won't allow it to be shared.  I clicked the "please review" button to see where that gets me.  If it's going to ding me every time then Google Drive isn't going to work.

**Reply by erichelgeson** [reply]
Thanks for flushing that out! Ya I have ditched most of google - Mega is an ok choice with a free tier. I think Github allows you to upload [LFS objects](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage) - maybe a 2nd repo with just the mac zip releases? I have linux on my desktop - I might try to write up the same if I can get it working.

**Reply by fadden** [reply]
It may be easiest just to add it to the release on github.

> Each file included in a release must be under 2 GiB. There is no limit on the total size of a release, nor bandwidth usage.

**Comment by erichelgeson** [comment]
Linux notes:

* Download and extract the `_fd` version of CiderPress
* Install wine 9 with the `wineWow` or `wine64` package (see your distro's package manager for correct commands)
* `$ wine boot` to setup default prefix in `~/.wine/`
* Install wine dependencies: `$ winetricks dotnetdesktop6 corefonts consolas droid` (droid fonts may not be required)
* Disable Avalon 3d acceleration to make menu items work:
`$ wine reg add "HKEY_CURRENT_USER\SOFTWARE\Microsoft\Avalon.Graphics" /v DisableHWAcceleration /t REG_DWORD /d 1 /f`

Start Ciderpress:
* `$ wine start CiderPress2.exe`
![Screenshot From 2025-01-22 09-10-47](https://github.com/user-attachments/assets/abaf692f-e64c-413d-9624-dcaa8fdeedf1)

**Reply by fadden** [reply]
Nifty... did you use the 32-bit or 64-bit version of cp2_fd?  (FWIW, I added an info-dump to CP2 v1.0.6.  Open Edit > Settings, check "Enable DEBUG menu", then DEBUG > System Info gives you a bunch of cryptic details about the .NET runtime and the environment it sees.)

The menu items didn't work for me on macOS with the 64-bit "engine", although you could usually get them to appear by using keyboard navigation instead of selecting them with the mouse.  And you can click on menu items, you just can't see them.  This actually reminds me of a bug I had with Mono and WinForms on Linux... https://github.com/mono/mono/issues/13619

Wine 10 was just released.  I don't know how long that'll take to show up in managed distributions though.

Does `wine --version` report `wine-9.0`?  That's what I get, and apt-get isn't seeing anything newer.  The last time I tried it things failed grandly, but maybe the configuration and winetricks stuff will help.

Thanks for working through this!

**Reply by erichelgeson** [reply]
Wine 9 yes - most distros wont change a major version until their next major release to not break users.

I used the 64 bit CP2 - It was recommended on a few sites to use the dotnet installer from winetricks vs a bundled (though in practice I don't think it made a difference.)

System Info:
```
CiderPress II GUI v1.0.6
+ DiskArc Library v1.0.6
+ Runtime: .NET 6.0.19 / win10-x64
  E_V=6.0.19
  E_OV=Microsoft Windows NT 10.0.19043.0
  E_OV_P=Win32NT
  E_64=True / True
  E_MACH="NIXOS" cpus=24
  E_CD="Z:\home\eric\Downloads\cp2_fd"
  RI_FD=.NET 6.0.19
  RI_OSA=X64
  RI_OSD=Microsoft Windows 10.0.19043
  RI_PA=X64
  RI_RI=win10-x64
  IsFreeBSD=False IsOSX=False IsLinux=False IsWindows=True
```

**Reply by fadden** [reply]
I've updated [WineNotes](https://github.com/fadden/CiderPress2/blob/main/WineNotes.md) with Linux info.  On my system I didn't actually need the registry tweak, but I did need to download an updated copy of winetricks.  As you noted, the "droid" fonts don't appear to be necessary.

I found some things online that indicated that dropping fonts into the emulated C:\Windows\Fonts should work, but it doesn't change anything for me.  So either I'm grabbing the wrong font files, or the WPF app is asking for a different font than I think it is.  (The easy way to see that the font is wrong is that the capital 'I's in "CiderPress II" should be sans-serif on the initial screen and about box.)

I uploaded a ZIP with the v1.0.6 macOS wrapper to [releases](https://github.com/fadden/CiderPress2/releases/tag/v1.0.6).

**Reply by erichelgeson** [reply]
Again thanks for taking my random notes and putting it into a doc :) Glad to see the release for GUI Mac users now - I always recommend it to Windows users when working with BlueSCSI images, now I can for Mac and Linux too!

The work around for gfx glitches might be due to me being on Nvidia drivers and/or Wayland.

**Comment by fadden** [comment]
It occurred to me that Segoe UI might be the wrong font if Wine is using really old GUI styles.  According to [this post](https://stackoverflow.com/a/4141880/294248), the font used in the WinXP days was Tahoma.  Adding it to the Linux installation with winetricks did improve the appearance:

Without:
![cp2-no-tahoma](https://github.com/user-attachments/assets/563818ff-fa67-4b38-9332-374333dddd33)

With:
![cp2-after-tahoma](https://github.com/user-attachments/assets/07b6571b-2dac-4ffe-9124-629432c9d326)

"Create new file archive" is now legible, though it's still blurry.

macOS already looked like the with-Tahoma output, and adding Tahoma had no noticeable effect.  Perhaps it's finding the font within macOS?

![cp2mac-no-tahoma](https://github.com/user-attachments/assets/c7b5ae14-579c-4d91-b1fd-b89b2580c3a4)

Curiously, resizing the window by a pixel or two (which causes the buttons to shift position) will fix the blurriness.  Maybe something in the font renderer is attempting sub-pixel rendering when it shouldn't.  I tried `winetricks settings fontsmooth=disable` but it didn't help.

At any rate, I've updated the Linux instructions to include Tahoma.

For reference, here's what it looks like on Windows 10:
![cp2-win](https://github.com/user-attachments/assets/2410384a-96fb-4ed8-b546-6dbfe610f509)
